qrexec-lib/unicode.c

```diff
@@ -101,25 +102,99 @@ static int validate_utf8_char(const uint8_t *untrusted_c) {
     return qubes_pure_code_point_safe_for_display(code_point) ? total_size : 0;
 }
 
-static size_t validate_path(const uint8_t *const untrusted_name, size_t allowed_leading_dotdot)
+// Statically assert that a statement is not reachable.
+//
+// At runtime, this is just abort(), but it comes with a neat trick:
+// if optimizations are on, CHECK_UNREACHABLE is defined, and the compiler
+// claims to be GNU-compatible, the compiler must prove that this is
+// unreachable.  Otherwise, it is a compile-time error.
+//
+// To enable static checking of this macro, pass CHECK_UNREACHABLE=1 to the
+// makefile or include it in the environment.
+#if defined __GNUC__ && defined __OPTIMIZE__ && defined CHECK_UNREACHABLE
+#define COMPILETIME_UNREACHABLE do {    \
+    extern void not_reachable(void)     \
+        __attribute__((                 \
+            error("Compiler could not prove that this statement is not reachable"), \
+            noreturn));                 \
+    not_reachable();                    \
+} while (0)
+#else
+#define COMPILETIME_UNREACHABLE do {    \
+    assert(0);                          \
+    abort();                            \
+} while (0)
+#endif
+
+// This is one of the trickiest, most security-critical functions in the
+// whole repository (opendir_safe() in unpack.c is the other).  It is critical
+// for preventing directory traversal attacks.  The code does use a chroot()
+// and a bind mount, but the bind mount is not always effective if mount
+// namespaces are in use, and the chroot can be bypassed (QSB-015).
+//
+// Preconditions:
+//
+// - untrusted_name is NUL-terminated.
+// - allowed_leading_dotdot is the maximum number of leading "../" sequences
+//   allowed.  Might be 0.
+// - allow_unsafe_chars is true if and only if unsafe characters should _not_
+//   be allowed.
```

Comment:
Isn't the comment the other way around? But also, there is no `allow_unsafe_chars`, but `flags` instead.